File: packages/server/src/utils/inline.ts
import { agentStreamEvent, type WorkflowEventData } from "@llamaindex/workflow";
import { type ChatMessage } from "llamaindex";
import { z } from "zod";

const INLINE_ANNOTATION_KEY = "annotation"; // the language key to detect inline annotation code in markdown

export const AnnotationSchema = z.object({
  type: z.string(),
  data: z.any(),
});

export type Annotation = z.infer<typeof AnnotationSchema>;

export function getInlineAnnotations(message: ChatMessage): Annotation[] {
  const markdownContent = getMessageMarkdownContent(message);

  const inlineAnnotations: Annotation[] = [];

  // Regex to match annotation code blocks
  // Matches ```annotation followed by content until closing ```
  const annotationRegex = new RegExp(
    `\`\`\`${INLINE_ANNOTATION_KEY}\\s*\\n([\\s\\S]*?)\\n\`\`\``,
    "g",
  );

  let match;
  while ((match = annotationRegex.exec(markdownContent)) !== null) {
    const jsonContent = match[1]?.trim();

    if (!jsonContent) {
      continue;
    }

    try {
      // Parse the JSON content
      const parsed = JSON.parse(jsonContent);

      // Validate against the annotation schema
      const validated = AnnotationSchema.parse(parsed);

      // Extract the artifact data
      inlineAnnotations.push(validated);
    } catch (error) {
      // Skip invalid annotations - they might be malformed JSON or invalid schema
      console.warn("Failed to parse annotation:", error);
    }
  }

  return inlineAnnotations;
}

/**
 * To append inline annotations to the stream, we need to wrap the annotation in a code block with the language key.
 * The language key is `annotation` and the code block is wrapped in backticks.
 *
 * \`\`\`annotation
 * \{
 *   "type": "artifact",
 *   "data": \{...\}
 * \}
 * \`\`\`
 */
export function toInlineAnnotation(item: unknown) {
  return `\n\`\`\`${INLINE_ANNOTATION_KEY}\n${JSON.stringify(item)}\n\`\`\`\n`;
}

export function toInlineAnnotationEvent(event: WorkflowEventData<unknown>) {
  return agentStreamEvent.with({
    delta: toInlineAnnotation(event.data),
    response: "",
    currentAgentName: "assistant",
    raw: event.data,
  });
}

function getMessageMarkdownContent(message: ChatMessage): string {
  let markdownContent = "";

  if (typeof message.content === "string") {
    markdownContent = message.content;
  } else {
    message.content.forEach((item) => {
      if (item.type === "text") {
        markdownContent += item.text;
      }
    });
  }

  return markdownContent;
}

File: python/llama-index-server/llama_index/server/utils/inline.py
import json
import re
from typing import Any, List

from pydantic import ValidationError

from llama_index.core.workflow.events import Event
from llama_index.server.models.chat import ChatAPIMessage
from llama_index.core.agent.workflow.workflow_events import AgentStream

INLINE_ANNOTATION_KEY = (
    "annotation"  # the language key to detect inline annotation code in markdown
)


def get_inline_annotations(message: ChatAPIMessage) -> List[Any]:
    """Extract inline annotations from a chat message."""
    markdown_content = message.content

    inline_annotations: List[Any] = []

    # Regex to match annotation code blocks
    # Matches ```annotation followed by content until closing ```
    annotation_regex = re.compile(
        rf"```{re.escape(INLINE_ANNOTATION_KEY)}\s*\n([\s\S]*?)\n```", re.MULTILINE
    )

    for match in annotation_regex.finditer(markdown_content):
        json_content = match.group(1).strip() if match.group(1) else None

        if not json_content:
            continue

        try:
            # Parse the JSON content
            parsed = json.loads(json_content)

            # Check for required fields in the parsed annotation
            if (
                not isinstance(parsed, dict)
                or "type" not in parsed
                or "data" not in parsed
            ):
                continue

            # Extract the annotation data
            inline_annotations.append(parsed)
        except (json.JSONDecodeError, ValidationError) as error:
            # Skip invalid annotations - they might be malformed JSON or invalid schema
            print(f"Failed to parse annotation: {error}")

    return inline_annotations


def to_inline_annotation(item: dict) -> str:
    """
    To append inline annotations to the stream, we need to wrap the annotation in a code block with the language key.
    The language key is `annotation` and the code block is wrapped in backticks.

    ```annotation
    {
      "type": "artifact",
      "data": {...}
    }
    ```
    """
    return f"\n```{INLINE_ANNOTATION_KEY}\n{json.dumps(item)}\n```\n"


def to_inline_annotation_event(event: Event) -> AgentStream:
    """
    Convert an event to an AgentStream with inline annotation format.
    """
    event_dict = event.model_dump()
    return AgentStream(
        delta=to_inline_annotation(event_dict),
        response="",
        current_agent_name="assistant",
        tool_calls=[],
        raw=event_dict,
    )
